File: programs/pixl/src/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct GameInitialized {
    pub game: Pubkey,
    pub authority: Pubkey,
    pub current_season_id: u32,
    pub timestamp: i64,
}

#[event]
pub struct SeasonStarted {
    pub game: Pubkey,
    pub season: Pubkey,
    pub season_id: u32,
    pub authority: Pubkey,
    pub title: String,
    pub start_time: i64,
    pub end_time: i64,
    pub timestamp: i64,
}

#[event]
pub struct CanvasInitialized {
    pub season: Pubkey,
    pub canvas: Pubkey,
    pub width: u16,
    pub height: u16,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct PlayerInitialized {
    pub player: Pubkey,
    pub wallet: Pubkey,
    pub max_energy: u8,
    pub energy_cooldown_seconds: u32,
    pub timestamp: i64,
}

#[event]
pub struct PlayerJoined {
    pub season: Pubkey,
    pub player: Pubkey,
    pub wallet: Pubkey,
    pub joined_at: i64,
    pub timestamp: i64,
}

#[event]
pub struct PixelPainted {
    pub player: Pubkey,
    pub season: Pubkey,
    pub x: u16,
    pub y: u16,
    pub old_color_index: u8,
    pub new_color_index: u8,
    pub timestamp: i64,
}

#[event]
pub struct SeasonEnded {
    pub game: Pubkey,
    pub season: Pubkey,
    pub season_id: u32,
    pub authority: Pubkey,
    pub ended_at: i64,
    pub timestamp: i64,
}

File: packages/sdk/events.ts
import { BorshCoder, EventParser, type Idl } from "@coral-xyz/anchor";
import type { PublicKey } from "@solana/web3.js";
import type BN from "bn.js";

export type PixlEventData = {
  GameInitialized: {
    game: PublicKey;
    authority: PublicKey;
    currentSeasonId: number;
    timestamp: BN;
  };
  SeasonStarted: {
    game: PublicKey;
    season: PublicKey;
    seasonId: number;
    authority: PublicKey;
    title: string;
    startTime: BN;
    endTime: BN;
    timestamp: BN;
  };
  CanvasInitialized: {
    season: PublicKey;
    canvas: PublicKey;
    width: number;
    height: number;
    authority: PublicKey;
    timestamp: BN;
  };
  PlayerInitialized: {
    player: PublicKey;
    wallet: PublicKey;
    maxEnergy: number;
    energyCooldownSeconds: number;
    timestamp: BN;
  };
  PlayerJoined: {
    season: PublicKey;
    player: PublicKey;
    wallet: PublicKey;
    joinedAt: BN;
    timestamp: BN;
  };
  PixelPainted: {
    player: PublicKey;
    season: PublicKey;
    x: number;
    y: number;
    oldColorIndex: number;
    newColorIndex: number;
    timestamp: BN;
  };
  SeasonEnded: {
    game: PublicKey;
    season: PublicKey;
    seasonId: number;
    authority: PublicKey;
    endedAt: BN;
    timestamp: BN;
  };
};

export type PixlEvent = {
  [K in keyof PixlEventData]: { name: K; data: PixlEventData[K] };
}[keyof PixlEventData];

export function createEventParser(idl: Idl, programId: PublicKey): EventParser {
  return new EventParser(programId, new BorshCoder(idl));
}

/** Parses all Pixl events emitted in a transaction's logs. */
export function parsePixlEvents(
  idl: Idl,
  programId: PublicKey,
  logs: string[]
): PixlEvent[] {
  const parser = createEventParser(idl, programId);
  const events: PixlEvent[] = [];
  for (const event of parser.parseLogs(logs)) {
    events.push({
      name: event.name,
      data: camelCaseKeys(event.data),
    } as PixlEvent);
  }
  return events;
}

/** Decodes a single base64-encoded event blob (e.g. from a CPI `emit_cpi!`). */
export function decodePixlEvent(idl: Idl, base64: string): PixlEvent | null {
  const decoded = new BorshCoder(idl).events.decode(base64);
  if (!decoded) return null;
  return { name: decoded.name, data: camelCaseKeys(decoded.data) } as PixlEvent;
}

// The borsh event coder returns snake_case field names; normalize to camelCase.
function camelCaseKeys(data: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data)) {
    const camel = key.replace(/_([a-z0-9])/g, (_, c: string) =>
      c.toUpperCase()
    );
    out[camel] = value;
  }
  return out;
}
